records/2024-01-22.reshaping-cicd-environments.md

```diff
@@ -0,0 +1,127 @@
+certified approvals by the change owners. Human interventions might be required
+at certain points to approve of changes or identify issues, potentially
+reverting bad rollouts. At the same time, having too many human intervention
+will lead to a higher number of mistakes.
+
+### Isolated environments for testing
+
+There are changes of certain nature such as infrastructure that leads to
+disruption of the environment itself and might require isolation in a separate
+environment.
+
+## Options
+
+This section lists suggestions made by the team with different considerations
+and goals. We'll first review the current CI/CD process.
+
+### Current process
+
+![Current CI/CD workflow](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/current-cicd.drawio.png?raw=true "Current CI/CD workflow")
+
+### Stronger responsibility for release changes on developers
+
+A lightweight changes that the process can include in the first place is to put
+more burden on the developers' shoulders on the changes approval step.
+Specifically, whatever the developer pushes to the staging environment is what
+will get released by the ops lead.
+
+This keeps the current environments as-is and does not make new automation
+additions. This is a status quo for most release di mensions except for the
+manual aspect.
+
+#### Pros
+
+- Incremental
+- Very cheap to implement
+- Lighten load on the ops lead role
+
+#### Cons
+
+- Higher risk factor of deploying to production (as it relies on multiple human interventions)
+
+### Create dedicated "Core" and "Dev" environments
+
+Currently the staging environment is used as an integration/development environment. This has been workng, but is not aligned with best practices, as it leaves staging in a non-production like state. With the introduction of the new "Dev" environment, it became possible to add an additional environment into the process that allows staging to always remain in the same state as production. 
+
+The new "Dev" envorinment as it exists today is not suitable for such a role because it is constantly being created/destroyed/altered by the Core team in order to test and develop infrastructure changes. It would not be reliable enough for the App team to develop against. 
+
+An additional environment could be created that is reserved for the App team's development and integration, and the existing "Dev" environment could be reclassified as the "Core" environment to identify it as being used by the Core team for infrastructure development.
+
+![Potential Core and Dev Environment Solution](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/2024-01-22.reshaping-cicd-environments/core-dev-env.png?raw=true "Potential Core and Dev Environment Solution")
+
+#### Pros
+  - Staging could be used to test the actual production release package and the associated process, which is currently not being tested.
```

Comment:
Can you add an example here? Specifically around the part - the associated process, which is currently not being tested. I am unsure what that is

Reply:
Per my understanding, we mean integration wise: we test and QA the individual parts but not the whole snapshot. For example, a backend change in the *notification-api* repository could lead to a bug with the front-end in a subtle manner, but we won't notice because we only run front-end tests with the *notification-admin* changes. Same can apply with the performance tests: we could run a subset of these against the release snapshots (but we don't).

Treating staging as dev also puts us in a position where the former environment cannot be released at certain point in time if we'd need to do a hot patch for an emergency release, because dev changes are in and these are not ready or tested.